### DISCUSSION TITLE
[RFC] Static Website

### DISCUSSION THREAD

**Discussion by fredericbarthelet** [getlift/lift#5] [Constructs]
The goal of this discussion is to get feedback on the "Static Website" component.

If you are new to Lift, a quick intro: it's a [Serverless](https://www.serverless.com/) plugin that can be installed via npm and enabled in any `serverless.yml` file.

Here is what we are planning so far.

## Use case

Deploying static websites and single-page applications, for example React, VueJS or Angular apps.

⚠️ Websites rendered server-side, i.e. with a backend language, will be covered by another component.

## Quick start

```yaml
# serverless.yml

# [...]

static-websites:

  landing:
    path: 'landing/dist'
```

The example above will create an S3 bucket to host the static website, and set up the CloudFront CDN.

*Note: this component takes about 5 minutes to deploy the first time.*

In the first version, the website will be deployed manually to S3 using the AWS CLI:

```bash
aws s3 sync landing/dist s3://bucket-name/ --delete
```

In the future, deployment will happen automatically during `serverless deploy`, including CDN cache invalidation.

## How it works

The component will deploy:

- an S3 bucket with OAI
- CloudFront CDN in front of S3 to provide HTTPS and caching at edge

### Cache invalidation

Quick intro on [how caching with CloudFront+S3 works](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/Expiration.html#ExpirationDownloadDist):

- By default, files are cached in CloudFront, not in browsers. Browsers send "If-Modified" requests.
    - Cache busting works by simply invalidating CloudFront.
- With `Cache-Control: max-age=3600` in S3 (on each object): CloudFront + browsers will cache for 1 hour.
    - Cache busting in CloudFront is possible with an invalidation. In the browser, the only solution is versioning.
- With `Cache-Control: max-age=60` and `Cache-Control: s-maxage=3600` in S3 (on each object): browsers will cache for 1 minute (max-age) and CloudFront for 1 hour (s-maxage).
    - Cache busting in CloudFront is possible with an invalidation. In the browser, the only solution is versioning.

Here are the use cases we plan on supporting:

1. I want a simple solution (no versioning) with edge caching.

    ```yaml
    static-websites:
      landing:
        path: 'public/landing'
    ```

    Lift deploys without `Cache-Control`. CloudFront will cache, but not browsers. On deployment, Lift creates a CloudFront invalidation that clears the CloudFront cache.

    After the deployment, every new page load should get the latest version.
    
    If a SPA was loaded in a browser before the deployment, every new asset loaded will be the new version.

    > Note: CloudFront invalidations are free up to 30 invalidations _per day_. Then it's 0.005$ per invalidation.

2. I want the most aggressive caching.

    ```yaml
    static-websites:
      landing:
        path: 'public/landing'
        browserCaching: true # could also be a duration, e.g. 3600
    ```
    
    In that scenario, website assets must contain a version number or hash in their filename ([for example in React](https://create-react-app.dev/docs/production-build/#static-file-caching)).

    Lift deploys with `Cache-Control=3600` **on all files except HTML**. CloudFront will cache everything, browsers will cache everything except HTML files.
    
    Why not HTML? So that browsers always reload the latest HTML version that points to the latest filenames.
    
    On deployment, Lift uploads the new files and creates a CloudFront invalidation which clears the CloudFront cache.
    
    New page loads will receive the latest HTML files, pointing to the latest assets versions.
    
    If a SPA was loaded in a browser before the deployment, it will keep trying to load old asset versions. This is why older versions are kept for 1 day in the S3 bucket. Lift clears old asset versions after 1 day.

### Custom domain

```yaml
static-websites:
  landing:
    domain: mywebsite.com
    # or (CloudFront accepts multiple domains)
    domain:
      - mywebsite.com
      - app.mywebsite.com
    # ARN of an ACM certificate for the domain, registered in us-east-1
    certificate: arn:aws:acm:us-east-1:123456615250:certificate/0a28e63d-d3a9-4578-9f8b-14347bfe8123
```

The configuration above will activate the custom domain mywebsite.com on CloudFront, using the provided HTTPS certificate.

The certificate must be registered in ACM in the `us-east-1` region before deploying.

After deploying, a CloudFront domain will be generated. Create a CNAME DNS entry for your domain that points to that `xxx.cloudfront.net` domain.

## Browser caching

```yaml
static-websites:
  landing:
    path: 'public/landing'
    browserCaching: true # could also be a duration, e.g. 3600
```

See the previous section about "Cache invalidation" for more details.

## Security

Ideally, we want some security headers to be present in responses.

How can we implement that? Lambda@Edge? → heavy solution

---

Feedback is welcome, just post a reply to this discussion 🙏

**Comment by M1ke** [comment]
> including CDN cache invalidation

I wonder could it also handle wiping the bucket content. E.g. for `create-react-app` each build generates custom JS files with a hash in the name relating to the build content. This means cache invalidation isn't needed, but also means that your bucket over time (especially if you're doing rapid development, CI/CD etc) fills up with lots of versions of very similar JS files, all but one of which are essentially old code. This could even be a security risk, e.g. you accidentally publish a key and then quickly upload a new build, but your file with the key is still there.

For good reason S3 doesn't make it easy to empty a bucket but for some application configs it'd be completely legitimate to have this as part of the framework.

**Reply by mnapoli** [reply]
Hey, that's a very good one. Cache invalidation is something central here.

In the "Cache invalidation" section I mention 2 options:

1. No versioning (or file hashes) are needed, because on every deploys we clear the CloudFront cache entirely. But: no browser caching. -> this will be the default option, so that it works easily out of the box.
2. Browser caching, which **requires versioning/hashing**. -> this will be optional.

It seems you are talking about option 2.

> This means cache invalidation isn't needed, but also means that your bucket over time fills up with lots of versions of very similar JS files

Yes! I've had that problem too. But I wonder if that problem will still exist.

Here is the workflow I have in mind (assuming *option 2*):

- `npm run build` (or whatever the command is) -> generates `index.html` + files with hashes
- `serverless deploy`
  - CloudFormation deployment happens
  - Then Lift runs `aws s3 sync <local> <bucket>` -> that **ADDS** new files to S3
  - While the S3 upload happens, old files are still available
  - Lift triggers a CloudFront invalidation
  - HTML files will now point to new versions of assets, old assets should never be loaded anymore
  - I think now is the perfect time to add `Cache-Control` headers to S3 objects -> EVERYTHING EXCEPT HTML files
  - When the 1st sync is done, Lift could run `aws s3 sync <local> <bucket> --delete` -> that will **DELETE** old versions

At this point, older versions are no longer in S3, we avoid accumulating endless versions.

Why add `Cache-Control` to everything except HTML files? Because HTML files are not versioned, so if they are cached in browsers then we can never bust the cache using versioning/hashing.

What do you think? Did I miss something? There could still be some race condition scenario in there, let me know if you can spot one. I think for most scenarios it's starting to be very robust.

**Reply by wysow** [reply]
I'm wondering why we need to trigger a CloudFront invalidation on *option 2*... Seems like CloudFront invalidation are expensive, right? So just updating the S3 bucket with the `--delete` option seems to do the trick and let CloudFront just invalidates old files whenever it wants...

**Reply by fredericbarthelet** [reply]
First 1000 CloudFront cache invalidation paths are free each month. Past this limit, you're right, it starts to get expensive with 0.005$ per invalidation path. Considering you'd like to invalidate only your `index.html` file in your bucket, it gives you a safe free pace of more than 30 releases per day (only one path per invalidation). If you don't invalidate the cache, then you might need to wait up to an hour (default cache value) to have your users eventually requesting your new index, referencing your new app js files version with its own hashes.

We could have best of both worlds:
- proceed with a cache invalidation in CloudFront when you haven't reach the end of your monthly free tier
- let CloudFront invalidate whenever cache expires for any additional deployment

But this mixed-behavior could be confusing when encountered for the first time

**Reply by andrea-cristaudo** [reply]
> What do you think? Did I miss something? There could still be some race condition scenario in there, let me know if you can spot one. I think for most scenarios it's starting to be very robust.

The scenario I'm thinking of is:
- user is using the static website during the deploy
- it's index.html is already loaded and the browser will not check for another copy until the next refresh
- assets are versioned but it's stuck on the previous deploy versions
- user starts to get unavailable resources -> the static website looks broken

I think it'll be better to keep at least one previous deployed version of everything on the bucket and delete and invalidate only versions older than how many versions you'd like to keep. Imho the best option to do something like this is generating some kind of state file on s3 bucket keep private.

What do you think?

**Reply by mnapoli** [reply]
> I'm wondering why we need to trigger a CloudFront invalidation on option 2...

We need to trigger it on HTML files so that they now point to the new versioned assets.

---

Regarding the cost of CloudFront invalidations, I don't think it's a huge problem personally. As @fredericbarthelet said, 30 free deploys per day is huge. And if you deploy more than that, you're probably in a company that can afford the cost.

No invalidating the cache means 1 hour deploy time, or even 1 day deploy time if we push up the default cache duration.

> But this mixed-behavior could be confusing when encountered for the first time

Yes, TBH small projects will never hit the limit (or get a huge bill). Large companies will very probably prefer the cost over deploys that take hours. If we want to be optimized for production, I'd really favor real invalidations.

We could document the cost though?

---

@andrea-cristaudo Ohhh this is what I love about working in the open, thank you! You've indeed spotted a scenario where this wouldn't work.

> I think it'll be better to keep at least one previous deployed version

Yes, and I would go even further than that. Let's say you do 3 deploys in 10 minutes, you need to keep more than 1 older version.

Maybe we should clear deployments that are more than a day old, that way we are really safe. And 1 day of history is very manageable in terms of bucket size/noise in the bucket.

Now how could we do that?

> Imho the best option to do something like this is generating some kind of state file on s3 bucket keep private.

Makes sense, but I would prefer to avoid creating extra resources (i.e. create a separate bucket), because it can be confusing for users.

I see 2 other possibilities:

- on deployment, Lift creates a S3 expiration rule on old files (from the previous version) so that they are removed in 24 hours -> not sure how easy/manageable/fast it is, might be worth a try
- on deployment, Lift removes files that are older than a day -> very simple if we manage to make it work (which is not sure yet)

In any case, that's good stuff to include in the design of the "perfect" solution.

**Reply by dhrrgn** [reply]
@mnapoli What about hooking into `aws:deploy:finalize`, and using the deployment artifacts in the deployment bucket to determine what assets to keep? It already handles pruning old artifacts (via `maxPreviousDeploymentArtifacts`).

Note: The `aws:deploy:finalize` hook gets invoked _before_ `aws:deploy:finalize:cleanup`, which is what cleans up the old deploy artifacts. 

You could use these artifacts for this in 1 of 2 ways:

1. Reading the date from the second-oldest (not oldest...see note above) deployment folder name, and deleting assets older than that.
2. After the upload, add the created s3 objects to the stack's output. Then during finalize, read in the list of assets from the oldest deployment's `compiled-cloudformation-template.json`, and delete those. 

Option 2 is safer, since Option 1 would also delete non-deployment assets if they are in the same bucket/path.

Either way, I think leveraging Serverless' existing deployment pruning makes the most sense (with an option to disable the pruning of course).

**Reply by mnapoli** [reply]
@dhrrgn thanks that's a good idea!

FYI I've updated the top post with the current state of the discussion.

I've also looked into service workers with the idea:

> if SPA can autorefresh when there are new versions, do we really need to keep old versions around?

So far my answer would be: yes, we need to keep old versions (for a day or two). Service workers are useful to show a popup to users "There's a new version available, reload?" but they don't force the reload. Also, it doesn't seem like a practice adopted by default, at least in React. So I don't think we can count on this.

**Reply by M1ke** [reply]
Just on that last point:

> with an option to disable the pruning of course

With option 2 you can have an option to disable, as the process _should_ be safe given it's aware of only specific files to delete.

With option 1 the deletion definitely needs to be opt-in as people can be using the bucket for anything else - it's bad practice but a general delete command can be [very dangerous](https://github.com/jedrichards/rsyncwrapper/issues/27)!

**Reply by mnapoli** [reply]
@M1ke this is a good point (sorry for not getting back sooner on that). The bucket for the static website is really a bucket containing _code artifacts_.

It isn't meant to store content at all, that's why it's 100% public and there are no backups enabled.

And to be clear: on every deployment made via `serverless deploy`, any extra file will be deleted. Just like when deploying to Netlify for example.

And yeah, we should 100% protect against mistakes. Besides documentation, do you see any way we can prevent accidental misuse? (i.e. no data uploaded in there)

**Reply by dhrrgn** [reply]
> And yeah, we should 100% protect against mistakes. Besides documentation, do you see any way we can prevent accidental misuse? (i.e. no data uploaded in there)

Since the website bucket is created and controlled via the plugin, I think the chance of misuse should be minimal. You'd have to go out of your way and try to do it.

That being said, the only thing I can think of is adding a check pre-deploy to make sure there are no non-deployed artifacts. This would slow down deploys though. It'd have to look something like:

- Add metadata to each deployed object via `--metadata` on the s3 sync call, like `--metadata "DeploymentArtifact=1`

Pre-deploy: 

1. List all objects (recursively in case there are >1000)
2. For each object, make a `GetObjectMetadata` request and if any objects are missing the `DeploymentArtifact` user metadata, abort the deploy with an error.

That sucks though, and adds a bunch of overhead for an extreme edge case (IMO). 

If someone _really_ needed to use the same bucket, they could add a policy to prevent deletion on their objects...which would probably break this component though since `emptyBucket` would error (not sure how s3 sync handles that...)

**Reply by mnapoli** [reply]
@dhrrgn that is a very good idea actually!

I am working on the deployment and I need to list all objects anyway (to upload only changed files). So that could be doable without overhead. I'll keep that in mind and probably try it later in the next weeks.

**Reply by dhrrgn** [reply]
@mnapoli Ah, very nice. It is annoying that you can't request user metadata in the list request. But, the `headObject` requests should be quick. Could make it faster by sending 10-20 async requests at a time and using `Promise.all` and rejecting if an unknown object. 

An alternative is similar to what I suggested previously: During deploy, add the keys for the uploaded objects to the CF output, or to a (non-public) manifest file. Then it's pretty trivial to detect changed/new/unknown files in the bucket. 🤷‍♂️

**Comment by M1ke** [comment]
Also whilst you're mentioning Lambda@Edge, worth noting that you can also do that for static site "neat" URLs, e.g. you have a directory structure like

```
public/
  - about/
  - - index.html
  - contact/
  - - index.html
  - index.html
```

So you want `domain.com/` to get the bottom `index.html` file, `domain.com/about` to get the one in that directory etc. You can't do that natively but Lambda@Edge on Cloudfront can rewrite it for you. I've documented that [here](https://m1ke.me/2018/12/deploy-a-secure-static-site-with-aws-terraform/) and also published a [Terraform module](https://github.com/M1ke/terraform-aws-secure-static/tree/master/modules/lambda-rewrite) to do that simply for those using TF to manage infrastructure.

**Reply by mnapoli** [reply]
`domain.com/about/` should work, do you mean that `domain.com/about` won't? (i.e. the difference is the trailing `/`?)

**Reply by M1ke** [reply]
I know that when I initially build my static site in 2018 I found that the structure I was used to using from static site generators like Jekyll which would turn `about.md` into `about/index.html` got a lot of not found errors. So I wrote that module, tested it, and it's been running ever since. Maybe it's not needed for every case, would make sense for Cloudfront to do that natively, but I'd need to check!

**Reply by M1ke** [reply]
@mnapoli seems we don't need Lambda@Edge for those rewrites any more!

https://aws.amazon.com/about-aws/whats-new/2021/05/cloudfront-functions/

**Reply by mnapoli** [reply]
@M1ke Yep I was waiting exactly for that feature ^^

**Reply by andrea-cristaudo** [reply]
https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/example-function-add-index.html

**Comment by wysow** [comment]
And just a thought but this feature should also work for static assets of a dynamic website I think...

**Reply by mnapoli** [reply]
Server-side websites (or backend websites) will be addressed in a separate component as this is a different use case that leads to a different solution. It will also let us provide options specific to server-side websites.

**Comment by andrea-cristaudo** [comment]
One common use case for static websites is redirecting www to non-www addresses or viceversa. With a web server is the easiest thing in the world but with aws-static-website-stack is a very complex task.
It should be very easy to implement in application level (something like what wordpress does) but it'd be very nice to get it in infrastructure level.

**Reply by mnapoli** [reply]
That's a very good point, TBH I never thought about how to solve that without Lambda@Edge (which is a PITA, if we can avoid it that's really great).

With S3 alone IIRC there's a technique with having 2 separate buckets (one for www, for for the root domain), but I'm not sure if it works with CloudFront. I'll have to research that.

**Reply by andrea-cristaudo** [reply]
Found out how to do without lambda@edge, it requires (for https support) one additional s3 bucket, one additional cloudfront instance and the dns record pointing to the cf instance.
I'll try to create a PR this week.

**Reply by mnapoli** [reply]
@andrea-cristaudo thanks. FYI don't bother with the PR yet, the codebase is very early and unstable.

Also I might have an idea on how to implement that in a simpler way!

**Reply by M1ke** [reply]
Yeah S3 is the way to go for www. redirects, [there's a Terraform example here](https://github.com/M1ke/terraform-aws-secure-static/blob/master/modules/static-website/cloudfront.tf)

**Reply by mnapoli** [reply]
CloudFront Functions are out, this might help here too.

**Comment by guillaumem-theodo** [comment]
About Security topic, adding default Security headers is a must have on this component. A set of default security headers could be a nice feature. Implementation using Lambda@Edge.

Something like:

```yaml
static-websites:
  landing:
    path: 'public/landing'
    securityHeaders: yes
```
or (adding additional headers)
```yaml
static-websites:
  landing:
    path: 'public/landing'
    securityHeaders:
      default: yes
      'X-XSS-Protection': '1; mode=block'
      ...
```

**Reply by mnapoli** [reply]
Right, this is something we have in mind. Ideally we wouldn't have to resort to Lambda@Edge because it's a huge added complexity, we'll look into other options.

**Reply by mnapoli** [reply]
CloudFront Functions are out and could help with that. It seems there will even be CloudFormation support with inline code, and CDK support as well: https://github.com/aws/aws-cdk/pull/14511/files#diff-3b0cd2595ecdf6f4a567197fcc6399233a11280b54a69b3730458f021fe6c933R616-R621

I'm waiting for CloudFormation support before trying out further (at the moment it's not documented yet).

**Reply by mnapoli** [reply]
CloudFormation is finally documented and supported: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cloudfront-function.html

I'll try that next week hopefully.

**Reply by andrea-cristaudo** [reply]
Example of adding security headers with CloudFront Functions: https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/example-function-add-security-headers.html

**Reply by mnapoli** [reply]
Thank you @andrea-cristaudo!

**Reply by mnapoli** [reply]
Finally managed to implement security headers using CloudFront Functions: https://github.com/getlift/lift/pull/21

Any review is welcome :)

**Comment by andrea-cristaudo** [comment]
Maybe would be useful to enable brotli compression as well?
It's not enough to set "compress: true" for cache behaviour but it's required to use a CachePolicy with a ParametersInCacheKeyAndForwardedToOrigin instance with EnableAcceptEncodingBrotli set to true. Here the link to the documentation: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-cloudfront-cachepolicy-parametersincachekeyandforwardedtoorigin.html

**Reply by mnapoli** [reply]
🤯 nice! Thanks a lot, will try to apply that next week!

**Reply by mnapoli** [reply]
I've opened https://github.com/getlift/lift/pull/22 to apply that recommendation, thank you!

Do you know any downside of switching to using "CachePolicy"? I haven't found any, but just in case.

**Comment by Pigius** [comment]
Out of curiosity, do you plan to support NextJS to the `static website construct` as well?

**Reply by t-richard** [reply]
@Pigius could you elaborate on your exact needs ?

If you use NextJS to generate a fully static website, this construct should work (I think you can do that by running `next export`).

If you use NextJS with SSR, then this static website construct would not be a fit. 

But **maybe** Server-side website construct would help here (see my last comment in https://github.com/getlift/lift/discussions/43#discussioncomment-896386).

**Reply by Pigius** [reply]
@t-richard ah I see, forgot about `next export` option. Will play around with it. Thanks for a quick response.

**Reply by Pigius** [reply]
@t-richard I played around with NextJS and static-website construct for a while and managed to do deployment after exporting to static HTML. However, not every project with nextJS can be exported this way (Not all functionalities are compatible with `next export`.).

However the deployment works, I even made a simple Proof of Concept: https://github.com/Pigius/nextjs-app-service. Thanks a lot for the help!

**Reply by mnapoli** [reply]
@Pigius very interesting, thanks for sharing!

I'd be interested to know what's missing to properly support NextJS (you mention missing features). Or maybe as @t-richard mentions, we would need backend capabilities (Lambda or Lambda@Edge) to fully support NextJS?

**Reply by Pigius** [reply]
You're welcome. I did not mention any missing features at all,  but I think what already @t-richard mentioned will make a job (SSR construct).

**Reply by t-richard** [reply]
@Pigius nice to know it worked !

These frameworks (Next, Nuxt, etc) are kinda limited when using only static generation. They are truly powerful when you enable server-side rendering (and that's their main purpose IMO).

In the case of SSR, then again, the Server-side website construct could be the place for that. I've only done SSR on Lambda with classic Lambdas, not Lambda@Edge and I think it is too restrictive for SSR to work properly.

**Reply by Pigius** [reply]
I see your point. Thanks!

**Comment by andrea-cristaudo** [comment]
Would be useful to let protect a static website with some kind of auth system. It would be great to provide staging versions of a website not public available.

Maybe with Cloudfront function could be achieved something like basic authentication with username/password credentials?

**Reply by mnapoli** [reply]
There seems to be interest in this (#78 mentions that too), so I've turned your comment into a new issue: #79.

Please add a 👍  on #79: we use these to prioritize new features.
